### ISSUE TITLE
Exception Task monitor

### ISSUE THREAD

**Issue by rrodriguez-consulting** [Cysharp/UniTask#86]
UnityException: get_dataPath can only be called from the main thread.
Constructors and field initializers will be executed from the loading thread when loading a scene.
Don't use this function in the constructor or field initializers, instead move initialization code to the Awake or Start function.

```
Cysharp.Threading.Tasks.Internal.DiagnosticsExtensions.AppendHyperLink (System.String path, System.String line) (at Assets/plugins/UniTask/Runtime/Internal/DiagnosticsExtensions.cs:242)
Cysharp.Threading.Tasks.Internal.DiagnosticsExtensions.CleanupAsyncStackTrace (System.Diagnostics.StackTrace stackTrace) (at Assets/plugins/UniTask/Runtime/Internal/DiagnosticsExtensions.cs:116)
Cysharp.Threading.Tasks.TaskTracker.TrackActiveTask (Cysharp.Threading.Tasks.IUniTaskSource task, System.Int32 skipFrame) (at Assets/plugins/UniTask/Runtime/Internal/TaskTracker.cs:73)
Cysharp.Threading.Tasks.UniTaskCompletionSource..ctor () (at Assets/plugins/UniTask/Runtime/UniTaskCompletionSource.cs:296)
Cysharp.Threading.Tasks.UniTask.FromException (System.Exception ex) (at Assets/plugins/UniTask/Runtime/UniTask.Factory.cs:37)
Cysharp.Threading.Tasks.CompilerServices.AsyncUniTaskMethodBuilder.get_Task () (at Assets/plugins/UniTask/Runtime/CompilerServices/AsyncUniTaskMethodBuilder.cs:39)
DataUploadService.ProcessData () (at <232b338c8d764c98a4446faa32d67655>:0)
Cysharp.Threading.Tasks.UniTask+<Run>d__49`1[T].MoveNext () (at Assets/plugins/UniTask/Runtime/UniTask.Run.cs:70)
UnityEngine.Debug:LogException(Exception)
Cysharp.Threading.Tasks.UniTaskScheduler:PublishUnobservedTaskException(Exception) (at Assets/plugins/UniTask/Runtime/UniTaskScheduler.cs:90)
Cysharp.Threading.Tasks.UniTaskCompletionSourceCore`1:ReportUnhandledError() (at Assets/plugins/UniTask/Runtime/UniTaskCompletionSource.cs:83)
Cysharp.Threading.Tasks.UniTaskCompletionSourceCore`1:Reset() (at Assets/plugins/UniTask/Runtime/UniTaskCompletionSource.cs:57)
Cysharp.Threading.Tasks.CompilerServices.AsyncUniTask`2:TryReturn() (at Assets/plugins/UniTask/Runtime/CompilerServices/StateMachineRunner.cs:249)
Cysharp.Threading.Tasks.CompilerServices.AsyncUniTask`2:Finalize() (at Assets/plugins/UniTask/Runtime/CompilerServices/StateMachineRunner.cs:321)
```

```csharp
static string AppendHyperLink(string path, string line)
        {
            var fi = new FileInfo(path);
            if (fi.Directory == null)
            {
                return fi.Name;
            }
            else
            {
                var fname = fi.FullName.Replace(Path.DirectorySeparatorChar, '/').Replace(Application.dataPath, "");
                var withAssetsPath = "Assets/" + fname;
                return "<a href=\"" + withAssetsPath + "\" line=\"" + line + "\">" + withAssetsPath + ":" + line + "</a>";
            }
        }
```

**Comment by KAW0** [comment]
This is expected behavior. Most Unity functionalities must be called from the main thread.
```
async static string AppendHyperLink(string path, string line)
    {
        var fi = new FileInfo(path);
        if (fi.Directory == null)
        {
            return fi.Name;
        }
        else
        {
            await UniTask.SwitchToMainThread();
            var dataPath= Application.dataPath;
            await UniTask.SwitchToTaskPool();
            var fname = fi.FullName.Replace(Path.DirectorySeparatorChar, '/').Replace(dataPath, "");
            var withAssetsPath = "Assets/" + fname;
            return "<a href="" + withAssetsPath + "" line="" + line + "">" + withAssetsPath + ":" + line + "";
        }
    }
```
This should work but I recommend to cache this value.

**Comment by rrodriguez-consulting** [comment]
it is called in Cysharp.Threading.Tasks.Internal.DiagnosticsExtensions

**Comment by neuecc** [comment]
Thank you for reporting.
It is called from finalizer thread and it is not unity main thread.
It's definitely a bug and we'll fix it.

**Comment by neuecc** [comment]
released in 2.0.15.
